src/db/mysql/batch.rs

```diff
@@ -172,49 +172,51 @@ pub fn do_append(
     db: &MysqlDb,
     batch_id: i64,
     user_id: HawkIdentifier,
-    collection_id: i32,
+    _collection_id: i32,
     bsos: Vec<params::PostCollectionBso>,
 ) -> Result<()> {
-    fn exist_idx(collection_id: i32, batch_id: i64, bso_id: &str) -> String {
+    fn exist_idx(user_id: u64, batch_id: i64, bso_id: &str) -> String {
+        // Construct something that matches the key for batch_upload_items
         format!(
-            "{collection_id}::{batch_id}::{bso_id}",
-            collection_id = collection_id,
+            "{batch_id}-{user_id}-{bso_id}",
             batch_id = batch_id,
+            user_id = user_id,
             bso_id = bso_id,
         )
     };
 
+    // It's possible for the list of items to contain a duplicate key entry.
+    // This means that we can't really call `ON DUPLICATE` here, because that's
+    // more about inserting one item at a time. (e.g. it works great if the
+    // values contain a key that's already in the database, less so if the
+    // the duplicate is in the value set we're inserting.
     #[derive(Debug, QueryableByName)]
+    #[table_name = "batch_upload_items"]
     struct ExistsResult {
-        #[sql_type = "Integer"]
-        batch: i32,
-        #[sql_type = "Integer"]
-        id: i32,
+        user_id: i64,
+        batch_id: i64,
+        id: String,
     };
 
-    /* This is not the most efficient way to do this using pure MySQL.
-       It would be more efficient to create an "INSERT ... ON DUPLICATE UPDATE"
-       query. This *does* match how spanner works, and may be easier to keep
-       in mind.
-    */
-
     let mut existing = HashSet::new();
-    let mut inserts: Vec<_> = Vec::new();
 
-    for item in sql_query("SELECT batch, id FROM batch_upload_items where userid=@userid;")
-        .bind::<BigInt, _>(user_id.legacy_id as i64)
-        .get_results::<ExistsResult>(&db.conn)?
+    // pre-load the "existing" table with any batched uploads that are already in the table.
```

Comment:
Nit, using the word table twice is confusing because the first time it's talking about a hashset and the second time it's talking about a sql table. Change the first use of table to hashset.